### DISCUSSION TITLE
Spot recovery strategy

### DISCUSSION THREAD

**Discussion by yuval-airis** [skypilot-org/skypilot#2567] [Ideas]
I saw that two StrategyExecutor for spot recovery are defined ([FailoverStrategyExecutor](https://github.com/skypilot-org/skypilot/blob/3cd931f18f2bd9d29270dd6963742131d31364c0/sky/spot/recovery_strategy.py#L366) and [EagerFailoverStrategyExecutor](https://github.com/skypilot-org/skypilot/blob/3cd931f18f2bd9d29270dd6963742131d31364c0/sky/spot/recovery_strategy.py#L455C7-L455C36) for the [sky.Resources](https://skypilot.readthedocs.io/en/latest/reference/api.html?highlight=preemption#resources) class.

The former tries to failover within the same cluster, and the latter prefers to failover to a machine in another cluster.

Can I also set some more complicated method of picking the right cluster? (e.g. the cheapest one / nearest one) Would that be achieved by implementing an executor of my own?

**Comment by concretevitamin** [comment]
> The former tries to failover within the same cluster, and the latter prefers to failover to a machine in another cluster. Can I also set some more complicated method of picking the right cluster? (e.g. the cheapest one / nearest one) 

Hey @yuval-airis. To clarify:
- [FailoverStrategyExecutor](https://github.com/skypilot-org/skypilot/blob/3cd931f18f2bd9d29270dd6963742131d31364c0/sky/spot/recovery_strategy.py#L366) does have a chance to launch a new cluster elsewhere.
- Both strategies use `sky.launch()` as a kernel to relaunch, and this function already optimizes for the cheapest and most available location (zone, region, cloud) within a given search space.

The two strategies differ in whether to back off from a preempted region quickly.

Yes, you could implement a new `StrategyExecutor` and put custom logic in `recover()`. Curious what policy are you thinking of?
